phoenix-core-client/src/main/java/org/apache/phoenix/compile/CreateTableCompiler.java

```diff
@@ -209,6 +247,137 @@ public MutationPlan compile(CreateTableStatement create) throws SQLException {
+        List<Integer> pkPositions = new ArrayList<>();
+        parentToBe.getPKColumns().forEach(tablePkColumn ->
+                tablePkPositions.add(tablePkColumn.getPosition()));
+        pkColumnsInWhere.forEach(pkColumn -> pkPositions.add(pkColumn.getPosition()));
+        Collections.sort(pkPositions);
+        // 2. If not multi tenant, view specification (WHERE clause) should start from the first PK
+        // column; otherwise, start from the second PK column
+        boolean isMultiTenant = parentToBe.isMultiTenant();
+        int firstPkPosition = pkPositions.get(0);
+        if (!isMultiTenant && !Objects.equals(firstPkPosition, tablePkPositions.get(0))
+                || isMultiTenant && !Objects.equals(firstPkPosition, tablePkPositions.get(1))) {
+            return ViewType.READ_ONLY;
+        }
+        // 3. Otherwise, PK column(s) should be in the order they are defined
+        if (!isPkColumnsInOrder(pkPositions, tablePkPositions, isMultiTenant)) {
+            return ViewType.READ_ONLY;
+        }
+
+        // 4. Check it uses the same set of PK column(s) as its sibling views' specification
+        byte[] parentTenantIdInBytes = parentToBe.getTenantId() != null
+                ? parentToBe.getTenantId().getBytes() : null;
+        byte[] parentSchemaNameInBytes = parentToBe.getSchemaName() != null
+                ? parentToBe.getSchemaName().getBytes() : null;
+        ConnectionQueryServices queryServices = connection.getQueryServices();
+        Configuration config = queryServices.getConfiguration();
+        byte[] systemChildLinkTable = SchemaUtil.isNamespaceMappingEnabled(null, config)
+                ? SYSTEM_CHILD_LINK_NAMESPACE_BYTES
+                : SYSTEM_CHILD_LINK_NAME_BYTES;
+        try (Table childLinkTable = queryServices.getTable(systemChildLinkTable)) {
+            List<PTable> legitimateSiblingViewList =
+                    ViewUtil.findAllDescendantViews(childLinkTable, config, parentTenantIdInBytes,
+                            parentSchemaNameInBytes, parentToBe.getTableName().getBytes(),
+                            HConstants.LATEST_TIMESTAMP, true).getFirst();
+            if (legitimateSiblingViewList.size() > 0) {
+                PTable siblingView = legitimateSiblingViewList.get(0);
+                Expression siblingViewWhere = getWhereFromView(connection, siblingView);
+                Set<PColumn> siblingViewPkColsInWhere = new HashSet<>();
+                if (siblingViewWhere != null) {
+                    ViewWhereExpressionValidatorVisitor siblingViewValidatorVisitor =
+                            new ViewWhereExpressionValidatorVisitor(parentToBe,
+                                    siblingViewPkColsInWhere, new HashSet<>());
+                    siblingViewWhere.accept(siblingViewValidatorVisitor);
+                }
+                if (!pkColumnsInWhere.equals(siblingViewPkColsInWhere)) {
+                    return ViewType.READ_ONLY;
```

Comment:
Every case where we return READ_ONLY view, can we add a log explaining why we need to create this as a read-only view?